nucleus/dataset_item.py

```diff
@@ -2,36 +2,135 @@
+from typing import Optional, Sequence, Dict, Any
+from enum import Enum
 
+from .annotation import is_local_path, Point3D
 from .constants import (
     DATASET_ITEM_ID_KEY,
     IMAGE_URL_KEY,
     METADATA_KEY,
     ORIGINAL_IMAGE_URL_KEY,
     REFERENCE_ID_KEY,
+    TYPE_KEY,
+    URL_KEY,
+    CAMERA_PARAMS_KEY,
+    POINTCLOUD_URL_KEY,
+    X_KEY,
+    Y_KEY,
+    Z_KEY,
+    W_KEY,
+    POSITION_KEY,
+    HEADING_KEY,
+    FX_KEY,
+    FY_KEY,
+    CX_KEY,
+    CY_KEY,
 )
 
 
 @dataclass
-class DatasetItem:
+class Quaternion:
+    x: float
+    y: float
+    z: float
+    w: float
 
-    image_location: str
+    @classmethod
+    def from_json(cls, payload: Dict[str, float]):
+        return cls(
+            payload[X_KEY], payload[Y_KEY], payload[Z_KEY], payload[W_KEY]
+        )
+
+    def to_payload(self) -> dict:
+        return {
+            X_KEY: self.x,
+            Y_KEY: self.y,
+            Z_KEY: self.z,
+            W_KEY: self.w,
+        }
+
+
+@dataclass
+class CameraParams:
+    position: Point3D
+    heading: Quaternion
+    fx: float
+    fy: float
+    cx: float
+    cy: float
+
+    @classmethod
+    def from_json(cls, payload: Dict[str, Any]):
+        return cls(
+            Point3D.from_json(payload[POSITION_KEY]),
+            Quaternion.from_json(payload[HEADING_KEY]),
+            payload[FX_KEY],
+            payload[FY_KEY],
+            payload[CX_KEY],
+            payload[CY_KEY],
+        )
+
+    def to_payload(self) -> dict:
+        return {
+            POSITION_KEY: self.position.to_payload(),
+            HEADING_KEY: self.heading.to_payload(),
+            FX_KEY: self.fx,
+            FY_KEY: self.fy,
+            CX_KEY: self.cx,
+            CY_KEY: self.cy,
+        }
+
+
+class DatasetItemType(Enum):
+    IMAGE = "image"
+    POINTCLOUD = "pointcloud"
+
+
+@dataclass  # pylint: disable=R0902
+class DatasetItem:  # pylint: disable=R0902
+    image_location: Optional[str] = None
     reference_id: Optional[str] = None
     item_id: Optional[str] = None
     metadata: Optional[dict] = None
+    pointcloud_location: Optional[str] = None
 
     def __post_init__(self):
         self.local = is_local_path(self.image_location)
+        assert bool(self.image_location) != bool(
+            self.pointcloud_location
+        ), "Must specify exactly one of the image_location, pointcloud_location parameters"
+        self.type = (
+            DatasetItemType.IMAGE
+            if self.image_location
+            else DatasetItemType.POINTCLOUD
+        )
+        camera_params = (
+            self.metadata.get(CAMERA_PARAMS_KEY, None)
+            if self.metadata
+            else None
+        )
+        self.camera_params = (
+            CameraParams.from_json(camera_params) if camera_params else None
+        )
 
     @classmethod
-    def from_json(cls, payload: dict):
-        url = payload.get(IMAGE_URL_KEY, "") or payload.get(
+    def from_json(cls, payload: dict, is_scene=False):
```

Comment:
This object should be serializable/deserializable independent of caller location.

Have to figure out a way to make this work without passing a parameter.

I think you can always do the is_scene logic but have to change the pointcloud url and image url to both default to none.
cls(
image_url = payload.get(IMAGE_URL_KEY, None) or payload.get(ORIGINAL_IMAGE_URL, None)
pointcloud_location = payload.get(POINTCLOUD_URL_KEY, None)....

Reply:
The serialization will work properly regardless of whether it's a scene, but the payload that's returned will be different. I believe the default to None is handled bc in the `__post_init__` we enforce that only one of image_url, pointcloud_url can be valid and the other is None

Reply:
The on-the-wire format should not depend on whether it is a scene. It's possible to have just one piece of logic that works for both, and it would be much cleaner.